### ISSUE TITLE
C2d initializeCompute after reuseOrder the previous provider fee is always expired

### ISSUE THREAD

**Issue by soonhuat** [oceanprotocol/multi-repo-issue#149]
**Describe the bug**
C2d initializeCompute dataset or algorithm transferTxId is always the asset order tx.

During initializeCompute provider will always use dataset or algorithm tx, regardless if user has been reuseOrder before.
https://github.com/oceanprotocol/provider/blob/c7f6b8cc3c1d17b00c4b7158aab60944d92ff0c4/ocean_provider/utils/provider_fees.py#L125
https://github.com/oceanprotocol/provider/blob/c7f6b8cc3c1d17b00c4b7158aab60944d92ff0c4/ocean_provider/utils/datatoken.py#L68
validate_order and datatoken_contract.events.ProviderFee().processReceipt is based on dataset or algorithm order tx, thus, the returned provider_fee_event_logs are from the asset order tx.
If reuseOrder been trigger before, should use the latest reuseOrder tx ProviderFee.

So if initializeCompute were trigger after exceeded computeEnv maxJobDuration will always has expired provider fee session.
When reuseOrder, the previous provider fee will always expired.

**To Reproduce**
1. Publish any C2D dataset and algorithm with timeout longer (maybe 1 week) than computeEnv maxJobDuration
2. Rerun the same pair of dataset and algorithm after maxJobDuration (default probably are 60min) but before dataset and algorithm expired.
4. handleComputeOrder will always goes to reuseOrder even-though were just reuseOrder 1min ago

**Contract Event**
https://github.com/oceanprotocol/contracts/blob/e60e5bc7a8d7780193b5d20341e25e750188e5d7/contracts/templates/ERC20TemplateEnterprise.sol#L532
<img width="186" alt="image" src="https://user-images.githubusercontent.com/7642607/169691672-110fdaa2-5538-4814-a4a5-f102b0a08146.png">
and 
https://github.com/oceanprotocol/contracts/blob/e60e5bc7a8d7780193b5d20341e25e750188e5d7/contracts/templates/ERC20TemplateEnterprise.sol#L413
<img width="288" alt="image" src="https://user-images.githubusercontent.com/7642607/169691694-620f43c8-6152-4081-be59-67c2564a02a1.png">
Seem like there is no relation to DT, not sure of subgraph could have event handlers to track the OrderReused or ProviderFee event, then passed from market as part of initializeCompute request.

**Comment by calina-c** [comment]
> During initializeCompute provider will always use dataset or algorithm tx, regardless if user has been reuseOrder before.

initializeCompute will use the "transferTxId" given to it by the user.
- If that order is still valid (both Provider fees validUntil and order timeout), then nothing needs to be regenerated. -> action: the user is good to go on their values
- If the Provider Fees have expired, then the same orderId will be returned to the user inside the "validOrder" key, but with new provider fees. -> action: the user needs to reuse the order. This will generate a new transaction id which should be replaced as the transferTxId for that asset.
- If the both provider fees and the order has expired, we return new provider fees and there will be no "validOrder" key. The user -> action: the user needs to order the asset again

As I understand it, you are describing a case where the user does not replace the "transferTxId" after reusing the order, but keeps the old txId from the initial order. Do I understand correctly?

**Comment by soonhuat** [comment]
@calina-c yes, the transferTxId is always the initial order tx, example https://rinkeby.etherscan.io/tx/0x439ab70d8008578461637fac8955ccf2dc53bfb0a7cf1d50439936fc808f99e1 and https://rinkeby.etherscan.io/tx/0xc97fa82cf6725f3ad97362ab01c2a632710517db40162942064c7a099743c9db

**Comment by soonhuat** [comment]
> @calina-c yes, the transferTxId is always the initial order tx, example https://rinkeby.etherscan.io/tx/0x439ab70d8008578461637fac8955ccf2dc53bfb0a7cf1d50439936fc808f99e1 and https://rinkeby.etherscan.io/tx/0xc97fa82cf6725f3ad97362ab01c2a632710517db40162942064c7a099743c9db

and the reuseOrder tx is: https://rinkeby.etherscan.io/tx/0xdb7c2d8eda09591454c890c09edf5345f3c958370a1e75290d56264af60eb272

**Comment by calina-c** [comment]
@soonhuat the current design expects that if an order was reused, the new tx should be sent. The user does have access to the new tx after orderReuse because they are required to make that transaction, and they can send the new tx.

I would advise against making any inferences in Provider, such as "guessing" the reused order tx based on a sent initial order.

**Comment by soonhuat** [comment]
@calina-c below is my sequence of action in market:
1. First time consume c2d.
2. Current setup of operator-engine computeEnv of 60min.
3. Buy dataset of 1 week timeout, https://rinkeby.etherscan.io/tx/0x439ab70d8008578461637fac8955ccf2dc53bfb0a7cf1d50439936fc808f99e1
4. Buy algorithm of 1 day timeout, https://rinkeby.etherscan.io/tx/0xff10a2696ba9b1b1afdd0da83a35bf2c5ef794f0524fc11782fcaad6dccea4a6
5. Compute job successfully started
6. 1st round completed without problem when start fresh new orders 
7. Wait after 4 hour, re-visit asset detail of the same compute job pair
8. initializeCompute checks on provider fee, response with new provider fees because provider fee expired (since only valid for 60min)
9. reuseOrder at transaction https://rinkeby.etherscan.io/tx/0xf38d6d786012a569a2c10f91b42a212df74d892d3c9fce663127ee96eaf41df1
10. Compute job successfully started after first time reuseOrder
11. after 13 min (by right provider fee should be still valid), refresh the same page.
12. initializeCompute checks on provider fee, response with new provider fees again, payload for initlizeCompute is using dataset order tx "transferTxId":"0x439ab70d8008578461637fac8955ccf2dc53bfb0a7cf1d50439936fc808f99e1"

I try query in subgraph with the reuseOrder tx, seem found nothing from query: orders, tokenTransactions
So market couldn't get back the reuseOrder tx in next session.

Based on above flow, is reuseOrder suppose to be single usage?
User would have to reuseOrder again when they refreshed the page even after 1min.

If by design, provider fee of reuseOrder are suppose to be single usage, then is as expected.
